README.md
# CS-HMI-Fibra
Fibra - Its like Libra but a fib! Although Libra is totally a fib too. This is a CS project, pls dont hurt me

*Note*: I will be keeping the server running for a few weeks or until this is graded. If it happens to go down, email me and I'll get it running again. However, all the information is in the Markdown docs.

## Sources
[bibliography.info](attribution/bibliography.info)

## Asset Credits
[assets.info](attribution/assets.info)

## Markdown Versions of Report Information
- [Fibra/Libra](info_db/fibra.md)
- [Marketplace](info_db/marketplace.md)
- [News](info_db/news.md)
- [Wallet](info_db/wallet.md)


info_db/news.md
**User Purchases**
The News section serves as a satirical method of displaying the privacy implications a system like Novi and Libra would cause. Although individual user purchase data would likely not be shown to other users, it's likely that it would be extensively sold to advertising agencies and statistical analysts. Though this would likely benefit the user experience, it would be a serious blow to individual privacy. If Libra proceeded to the point where it replaced or at least overtook conventional currency, it would mean that most of the average consumer's purchasing history would be easily viewable to Facebook, the Libra Association, and their clients. As the system is centralized, this data would also be subsceptible to leaks and cyberattacks, possibly leaking damaging information about Libra users to the general public.

**Breaking News - Fibra/Libra Association**
A fact that has not yet been addressed is that the Libra Association has complete control over the conversion rate between Libra and various currencies. While this might not be too much of an issue, there is very little stopping them from manipulating prices to keep people within the system, or make themselves some easy money. The whitepaper indicates that this won't happen, but that is (for all intents and purposes) little better than a spoken arrangement (Libra White Paper).

connbase.py
from fastapi import FastAPI, Response, status
from fastapi_utils.tasks import repeat_every
import time, json
import uvicorn

class App:
    def __init__(
        self,
        user_management=True,
        session_timeout=30,
        user_cache='users.json'
        ):
        self.app = FastAPI()
        self.connections = {}
        self.users = {}

        # Config options
        self.user_management = user_management
        self.session_timeout = session_timeout
        self.user_cache = user_cache

        # Request events
        # Connection requests
        @self.app.post('/connections/new/')
        async def new_connection(fingerprint: str):
            if fingerprint in self.connections.keys():
                self.connections[fingerprint]['last_update'] = time.time()
            else:
                self.connections[fingerprint] = self.create_connection(fingerprint)
            return self.connections[fingerprint]
        @self.app.get('/connections/self/')
        async def get_connection(fingerprint: str, response: Response):
            if not self.check_fp(fingerprint):
                response.status_code = status.HTTP_404_NOT_FOUND
                return {
                    'connection':None,
                    'user':None
                }
            self.connections[fingerprint]['last_update'] = time.time()
            
            if self.connections[fingerprint]['current_user']:
                usr = self.users[self.connections[fingerprint]['current_user']]
            else:
                usr = None
            conn = self.connections[fingerprint].copy()
            self.connections[fingerprint]['update'] = False
            return {
                'connection':conn,
                'user':usr
            }
        
        # Interval events
        @self.app.on_event('startup')
        @repeat_every(seconds=120)
        async def check_inactive_connections():
            nc = {}
            for conn in self.connections.values():
                if conn['last_update']+self.session_timeout >= time.time():
                    nc[conn['fingerprint']] = conn.copy()
            self.connections = nc.copy()


    
    def run(self,host='localhost',port=5000,log_level='info'):
        uvicorn.run(self.app,host=host,port=port,log_level=log_level,access_log=False)
    
    def check_fp(self,fingerprint):
        if not fingerprint in self.connections.keys():
            self.cache_all()
            self.users = self.load_cache()
            found = False
            for u in self.users.values():
                if u['owner'] == fingerprint:
                    self.connections[fingerprint] = self.create_connection(fingerprint)
                    found = True
                    break
            return found
        return True
    
    def update(self,fingerprint):
        self.connections[fingerprint]['update'] = True
        self.cache_all()
                
    
    def create_connection(self,fingerprint):
        return {
            'fingerprint':fingerprint,
            'last_update':time.time(),
            'current_user':None,
            'creation':time.time(),
            'update':True
        }
    
    # User caching ops
    def cache(self,user):
        with open(self.user_cache,'r') as f:
            old_cache = json.load(f)
        old_cache[user] = self.users[user]
        with open(self.user_cache,'w') as f:
            json.dump(old_cache,f)
    def get_cached(self,user):
        with open(self.user_cache,'r') as f:
            old_cache = json.load(f)
        return old_cache[user]
    def del_cached(self,user):
        with open(self.user_cache,'r') as f:
            old_cache = json.load(f)
        del old_cache[user]
        with open(self.user_cache,'w') as f:
            json.dump(old_cache,f)
    def load_cache(self):
        with open(self.user_cache,'r') as f:
            return json.load(f)
    def cache_all(self):
        with open(self.user_cache,'r') as f:
            old_cache = json.load(f)
        for user in self.users.keys():
            old_cache[user] = self.users[user]
        with open(self.user_cache,'w') as f:
            json.dump(old_cache,f)

main.py
from connbase import *
from hashlib import new, sha256
import os
from fastapi.responses import FileResponse
from markdown2 import markdown
import random
from pydantic import BaseModel

EXTRAS = [
    'break-on-newline',
    'cuddled-lists',
    'header-ids',
    'nofollow',
    'strike',
    'target-blank-links',
    'tables'
]

MAX_BOT_ITEMS = 3

ADJS = ['Shiny','Weird','Strange','Heavy','Light','Potato-Shaped','Really Expensive','Small','Large','Long','Stubby']
COLORS = ['Red','Blue','Green','Purple','Yellow','Orange','Dark Blue','Black','Gray','Light Gray','White']


class MainApp(App):
    def __init__(
            self,
            user_management=True,
            session_timeout=30,
            user_cache='users.json',
            c_rate=1.5
        ):
        super().__init__()
        self.conversion_rate = c_rate
        self.news = []
        self.purchases = []
        self.last_con_change = 0

    def create_connection(self, fingerprint):
        conn = super().create_connection(fingerprint)
        self.cache_all()
        self.users = self.load_cache()
        usr = None
        for u in self.users.keys():
            if self.users[u]['owner'] == fingerprint:
                usr = u
                break
        if usr == None:
            usr = sha256(fingerprint.encode('utf-8')).hexdigest()
            self.users[usr] = {
                'id':usr,
                'owner':fingerprint,
                'funds':50,
                'dollars':50,
                'inventory':[],
                'name':'Citizen #'+str(1+len(list(self.users.keys()))),
                'events':[]
            }
            self.users[usr]['events'].append({
                'type':'toast',
                'text':'Welcome!'
            })
            conn['update'] = True
        conn['current_user'] = usr
        self.cache_all()
        return conn
    
    def broadcast(self,event,user=None):
        for u in self.users.keys():
            if u == user or user == None:
                self.users[u]['events'].append(event)
                self.update(self.users[u]['owner'])

app = MainApp()
if not os.path.exists('users.json'):
    with open('users.json','w') as f:
        f.write('{}')

for i in range(5):
    usr = sha256(str(random.random()).encode('utf-8')).hexdigest()
    app.users[usr] = {
        'id':usr,
        'owner':'system',
        'funds':50,
        'dollars':0,
        'inventory':[],
        'name':'Bot #'+str(1+len(list(app.users.keys()))),
        'events':[]
    }
app.cache_all()
app.users = app.load_cache()

@app.app.get('/server/')
async def get_server():
    return {
        'conversion_rate':app.conversion_rate,
        'news':app.news
    }

@app.app.post('/user/name/')
async def change_name(fingerprint: str,name: str, response: Response):
    if not app.check_fp(fingerprint):
        response.status_code = status.HTTP_404_NOT_FOUND
        return
    app.users[app.connections[fingerprint]['current_user']]['name'] = name
    app.update(fingerprint)
    return

@app.app.get('/user/all/')
async def get_users(response: Response):
    app.cache_all()
    app.users = app.load_cache()
    r_users = {}
    for u in app.users.keys():
        for_sale = []
        for item in app.users[u]['inventory']:
            if item['for_sale']:
                for_sale.append({
                    'id':item['id'],
                    'name':item['name'],
                    'seller':u,
                    'seller_name':app.users[u]['name'],
                    'price':item['price']
                })
        r_users[u] = {
            'name':app.users[u]['name'],
            'for_sale':for_sale
        }
    return r_users

@app.app.get('/info/')
async def get_info(infoName: str, response: Response):
    if infoName+'.md' in os.listdir('info_db'):
        with open(os.path.join('info_db',infoName+'.md'),'r') as imd:
            return {
                'content':markdown(imd.read(),extras=EXTRAS)
            }
    else:
        return {
            'content':'<div>No content provided for INFO_'+infoName+'. Please contact system administrator.</div>'
        }

@app.app.post('/user/events/dequeue')
async def dequeue(fingerprint: str, response: Response):
    if not app.check_fp(fingerprint):
        response.status_code = status.HTTP_404_NOT_FOUND
        return
    if len(app.users[app.connections[fingerprint]['current_user']]['events']) > 0:
        del app.users[app.connections[fingerprint]['current_user']]['events'][0]
        app.update(fingerprint)
        return
    else:
        response.status_code = status.HTTP_204_NO_CONTENT
        return

class PurchaseModel(BaseModel):
    fingerprint: str
    name: str
    price: int
    id: str
    uid: str

@app.app.post('/user/items/purchase/')
async def purchase(fingerprint: str, name: str, price: str, id: str, uid: str, response: Response):
    price = int(price)
    if not app.check_fp(fingerprint):
        response.status_code = status.HTTP_404_NOT_FOUND
        return
    found = False
    for u in app.users:
        if u == uid:
            found = True
            break
    if not found:
        response.status_code = status.HTTP_404_NOT_FOUND
        return
    if price > app.users[app.connections[fingerprint]['current_user']]['funds']:
        response.status_code = status.HTTP_403_FORBIDDEN
        return
    app.users[app.connections[fingerprint]['current_user']]['funds'] -= price
    app.users[app.connections[fingerprint]['current_user']]['inventory'].append({
        'id':sha256(str(time.time()*random.random()).encode('utf-8')).hexdigest(),
        'name':name,
        'price':price,
        'for_sale':False
    })
    app.purchases.append({
        'username':app.users[app.connections[fingerprint]['current_user']]['name'],
        'item_name':name,
        'price':price
    })
    app.users[uid]['funds'] += price
    for i in range(len(app.users[uid]['inventory'])):
        if app.users[uid]['inventory'][i]['id'] == id:
            del app.users[uid]['inventory'][i]
            break
    
    app.update(fingerprint)
    return

@app.app.post('/user/items/edit/{item}/')
async def edit_item(item,fingerprint: str, key: str, value: str, response: Response):
    if not app.check_fp(fingerprint):
        response.status_code = status.HTTP_404_NOT_FOUND
        return
    if key == 'name':
        val = str(value)
    elif key == 'price':
        val = int(value)
    elif key == 'for_sale':
        if value == 'true':
            val = True
        else:
            val = False
    else:
        response.status_code = status.HTTP_405_METHOD_NOT_ALLOWED
        return
    for i in range(len(app.users[app.connections[fingerprint]['current_user']]['inventory'])):
        if app.users[app.connections[fingerprint]['current_user']]['inventory'][i]['id'] == item:
            app.users[app.connections[fingerprint]['current_user']]['inventory'][i][key] = val
            app.update(fingerprint)
            return
    response.status_code = status.HTTP_404_NOT_FOUND
    return

@app.app.post('/user/funds/edit/')
async def edit_funds(fingerprint: str, fibra: float, dollars: float, response: Response):
    if not app.check_fp(fingerprint):
        response.status_code = status.HTTP_404_NOT_FOUND
        return
    app.users[app.connections[fingerprint]['current_user']]['funds'] = fibra
    app.users[app.connections[fingerprint]['current_user']]['dollars'] = dollars
    return

@app.app.on_event('startup')
@repeat_every(seconds=5)
async def simloop():
    app.cache_all()
    app.users = app.load_cache()
    with open('items.json','r') as f:
        items = json.load(f)
    newitem = False
    for u in app.users.keys():
        if app.users[u]['owner'] == 'system' and random.randint(0,100) > 95 and len(app.users[u]['inventory']) < MAX_BOT_ITEMS:
            item = random.choice(items)
            app.users[u]['inventory'].append({
                'id':sha256(str(time.time()*random.random()).encode('utf-8')).hexdigest(),
                'name':str(item['name']).format(
                        adj=random.choice(ADJS),
                        adj2=random.choice(ADJS),
                        adj3=random.choice(ADJS),
                        color=random.choice(COLORS),
                        color2=random.choice(COLORS),
                        color3=random.choice(COLORS)
                    ),
                'price':item['price']+max([1,random.randint(-10,10)]),
                'for_sale':True
            })
            newitem = True
        if app.users[u]['owner'] == 'system' and len(app.users[u]['inventory']) < MAX_BOT_ITEMS:
            usr = {'owner':'system'}
            c = 0
            while usr['owner'] == 'system' and c < 2*len(app.users.values()):
                usr = random.choice(list(app.users.values()))
                c+=1
            if not usr['owner'] == 'system':
                old_inv = []
                for i in range(len(usr['inventory'])):
                    if usr['inventory'][i]['for_sale'] and random.random() > float(int(app.users[u]['funds']) / int(usr['inventory'][i]['price']*1.5)):
                        app.users[usr['id']]['funds'] += usr['inventory'][i]['price']
                        app.users[u]['funds'] -= usr['inventory'][i]['price']
                        usr['inventory'][i]['for_sale'] = True
                        usr['inventory'][i]['price'] += random.randint(-2,15);
                        app.users[u]['inventory'].append(usr['inventory'][i].copy())
                        app.broadcast({
                            'type':'toast',
                            'text':app.users[u]['name'] + ' has bought your ' + usr['inventory'][i]['name'] + '.'
                        },user=usr['owner'])
                    else:
                        old_inv.append(usr['inventory'][i].copy())
                app.users[usr['id']]['inventory'] = old_inv[:]
    if newitem:
        for i in app.users.keys():
            app.update(app.users[i]['owner'])
    
    with open('news_template.json','r') as f:
        ntemp = json.load(f)
    npurch = []
    for i in range(len(app.purchases)):
        if random.random() > 0.3:
            app.news.append(str(random.choice(ntemp)).format(name=app.purchases[i]['username'],item=app.purchases[i]['item_name'],price='≋'+str(app.purchases[i]['price'])))
        else:
            npurch.append(app.purchases[i].copy())
    app.purchases = npurch[:]

    if app.last_con_change+120 < time.time() and random.random() >= 0.98:
        app.conversion_rate += round((random.choice([1,-1]) * random.random())*0.4,2)
        app.conversion_rate = max([0.25,app.conversion_rate])
        app.last_con_change = time.time()
        app.news.append('BREAKING: Fibra Association arbitrarily sets conversion rate to '+str(app.conversion_rate)+' dollars per Fibra!')
            

# Load all static files in web/
@app.app.get('/',include_in_schema=False)
async def web_main():
    return FileResponse(os.path.join('web','index.html'))

static = os.walk('web')
files = {}
for folder in list(static):
    for f in folder[2]:
        files['/'.join(folder[0].split(os.sep))+'/'+f] = os.path.abspath(folder[0] + os.sep + f)

for f in files.keys():
    code = '\n'.join([
        '@app.app.get("/'+f.split('web/',maxsplit=1)[1]+'",include_in_schema=False)',
        'async def web_'+f.split('/')[len(f.split('/'))-1].replace('.','_').replace('-','_').replace(' ','_').replace('\'','').replace('"','')+'():',
        '\treturn FileResponse(r"'+files[f]+'")'
    ])
    exec(
        code,
        globals(),
        locals()
    )

app.run()